File: src/evolveguard/fixtures.ts
import * as fs from 'node:fs';
import { z } from 'zod';
import type { Fixture } from './types';
import { EvolveGuardError } from './errors';

const expectedToolCallSchema = z.object({
  tool: z.string().min(1),
  scopeMatches: z.string().optional(),
});

const fixtureSchema = z.object({
  id: z.string().min(1),
  prompt: z.string().min(1),
  expectedToolCalls: z.array(expectedToolCallSchema).optional(),
});

const fixturesFileSchema = z.array(fixtureSchema).min(1);

/** Reads and validates a user-supplied fixtures.json against the documented schema. */
export function loadFixtures(fixturesPath: string): Fixture[] {
  let raw: string;
  try {
    raw = fs.readFileSync(fixturesPath, 'utf8');
  } catch (err) {
    throw new EvolveGuardError(
      `Could not read fixtures file at "${fixturesPath}".`,
      err instanceof Error ? err.message : String(err),
      'Pass a valid path to a fixtures JSON file with --fixtures.'
    );
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    throw new EvolveGuardError(
      `Fixtures file at "${fixturesPath}" is not valid JSON.`,
      err instanceof Error ? err.message : String(err),
      'Fix the JSON syntax, e.g. run it through `node -e "JSON.parse(require(\'fs\').readFileSync(process.argv[1]))"`.'
    );
  }

  const result = fixturesFileSchema.safeParse(parsed);
  if (!result.success) {
    throw new EvolveGuardError(
      `Fixtures file at "${fixturesPath}" does not match the expected schema.`,
      result.error.issues.map((i) => `${i.path.join('.')}: ${i.message}`).join('; '),
      'A fixtures file is a non-empty JSON array of {"id": string, "prompt": string, "expectedToolCalls"?: [{"tool": string, "scopeMatches"?: string}]}.'
    );
  }

  const ids = new Set<string>();
  for (const fixture of result.data) {
    if (ids.has(fixture.id)) {
      throw new EvolveGuardError(
        `Duplicate fixture id "${fixture.id}" in "${fixturesPath}".`,
        'Fixture ids must be unique within a fixtures file -- they are used to match baseline entries on replay.',
        'Give each fixture a unique id.'
      );
    }
    ids.add(fixture.id);
  }

  return result.data;
}

File: python/src/evolveguard/fixtures.py
"""
Reads and validates a user-supplied fixtures.json against the documented
schema: a non-empty JSON array of
{"id": string, "prompt": string, "expectedToolCalls"?: [{"tool": string, "scopeMatches"?: string}]}.
Ported from src/evolveguard/fixtures.ts (which uses zod; this port validates
by hand to avoid a schema-validation dependency).
"""
from __future__ import annotations

import json
from typing import Any, List

from .errors import EvolveGuardError
from .types import ExpectedToolCall, Fixture

_SCHEMA_HINT = (
    'A fixtures file is a non-empty JSON array of {"id": string, "prompt": string, '
    '"expectedToolCalls"?: [{"tool": string, "scopeMatches"?: string}]}.'
)


def _schema_error(fixtures_path: str, reason: str) -> EvolveGuardError:
    return EvolveGuardError(
        f'Fixtures file at "{fixtures_path}" does not match the expected schema.',
        reason,
        _SCHEMA_HINT,
    )


def _parse_expected_tool_call(raw: Any, index: int) -> ExpectedToolCall:
    if not isinstance(raw, dict):
        raise ValueError(f"expectedToolCalls[{index}]: expected an object")
    tool = raw.get("tool")
    if not isinstance(tool, str) or not tool:
        raise ValueError(f"expectedToolCalls[{index}].tool: expected a non-empty string")
    scope_matches = raw.get("scopeMatches")
    if scope_matches is not None and not isinstance(scope_matches, str):
        raise ValueError(f"expectedToolCalls[{index}].scopeMatches: expected a string")
    return ExpectedToolCall(tool=tool, scope_matches=scope_matches)


def _parse_fixture(raw: Any, index: int) -> Fixture:
    if not isinstance(raw, dict):
        raise ValueError(f"[{index}]: expected an object")
    fid = raw.get("id")
    if not isinstance(fid, str) or not fid:
        raise ValueError(f"[{index}].id: expected a non-empty string")
    prompt = raw.get("prompt")
    if not isinstance(prompt, str) or not prompt:
        raise ValueError(f"[{index}].prompt: expected a non-empty string")

    raw_expected = raw.get("expectedToolCalls")
    if raw_expected is None:
        expected = None
    else:
        if not isinstance(raw_expected, list):
            raise ValueError(f"[{index}].expectedToolCalls: expected an array")
        expected = [
            _parse_expected_tool_call(item, i) for i, item in enumerate(raw_expected)
        ]

    return Fixture(id=fid, prompt=prompt, expected_tool_calls=expected)


def load_fixtures(fixtures_path: str) -> List[Fixture]:
    try:
        with open(fixtures_path, "r", encoding="utf-8") as fh:
            raw_text = fh.read()
    except OSError as err:
        raise EvolveGuardError(
            f'Could not read fixtures file at "{fixtures_path}".',
            str(err),
            "Pass a valid path to a fixtures JSON file with --fixtures.",
        ) from err

    try:
        parsed = json.loads(raw_text)
    except json.JSONDecodeError as err:
        raise EvolveGuardError(
            f'Fixtures file at "{fixtures_path}" is not valid JSON.',
            str(err),
            "Fix the JSON syntax, e.g. run it through `python3 -m json.tool <file>`.",
        ) from err

    if not isinstance(parsed, list) or len(parsed) == 0:
        raise _schema_error(
            fixtures_path, "expected a non-empty array, got " + type(parsed).__name__
        )

    try:
        fixtures = [_parse_fixture(item, i) for i, item in enumerate(parsed)]
    except ValueError as err:
        raise _schema_error(fixtures_path, str(err)) from err

    ids = set()
    for fixture in fixtures:
        if fixture.id in ids:
            raise EvolveGuardError(
                f'Duplicate fixture id "{fixture.id}" in "{fixtures_path}".',
                "Fixture ids must be unique within a fixtures file -- they are used to "
                "match baseline entries on replay.",
                "Give each fixture a unique id.",
            )
        ids.add(fixture.id)

    return fixtures
